### PR TITLE
refactor: team pages state logic

### PR DESCRIPTION
**Pull request by JoaoSaIvador**
Relates to:
- #1248 

Fixes:
- #1248 

## Proposed Changes

- Consolidate repeated code into reusable functions or modules
- Create a more structured and modular approach to state management
- Use a consistent naming convention for variables and functions to improve clarity and readability
- Fewer bugs and easier debugging, as there will be less repeated and inconsistent code
- Faster development time, as developers will be able to more easily reuse existing code
- Improved scalability and maintainability, as the updated architecture will be more organized and easier to maintain over time

This pull request closes:
- #1248

### CODE REVIEW

File: frontend/src/components/Teams/Team/ListMembers/ListMembers.tsx [hunk 1]
```diff
@@ -25,7 +25,7 @@ const ListMembers = ({ isOpen, setIsOpen, isTeamPage }: Props) => {
 
   const router = useRouter();
```
Comment (nunocaseiro) [inline frontend/src/components/Teams/Team/ListMembers/ListMembers.tsx:26]: I know you didn't change this line but you can do this:

```suggestion
 const {
    query: { teamId },
  } = useRouter();
```

File: frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx [hunk 2]
```diff
@@ -1,57 +1,82 @@
...
 
-const TeamItem = React.memo<TeamItemProps>(({ team, isTeamPage }) => {
-  const { data: session } = useSession();
+const TeamItem = React.memo(({ team }: TeamItemProps) => {
+  const { id, users: teamUsers, name } = team;
+
+  // CHECK: the session data could be passed as props,
+  // since it's the same for all items
+  const { userId, isSAdmin } = useCurrentSession();
   const router = useRouter();
```
Comment (nunocaseiro) [inline frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx:33]:
```suggestion
  const {pathname} = useRouter();
```

File: frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx [hunk 3]
```diff
@@ -1,57 +1,82 @@
...
+  const isTeamPage = router.pathname.includes('teams');
```
Comment (nunocaseiro) [inline frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx:34]:
```suggestion
  const isTeamPage = pathname.includes('teams');
```

File: frontend/src/components/Teams/CreateTeam/CreateTeam.tsx [hunk 4]
```diff
@@ -11,50 +11,30 @@ import Button from '@/components/Primitives/Inputs/Button/Button';
...
 import TipBar from '@/components/Primitives/Layout/TipBar/TipBar';
 import CreateHeader from '@/components/Primitives/Layout/CreateHeader/CreateHeader';
 import CreateFooter from '@/components/Primitives/Layout/CreateFooter/CreateFooter';
+import useCurrentSession from '@/hooks/useCurrentSession';
+import { CREATE_TEAM_TIPS } from '@/utils/tips';
 import ListMembers from '../Team/ListMembers/ListMembers';
 
 const CreateTeam = () => {
+  const { userId } = useCurrentSession();
   const router = useRouter();
```
Comment (nunocaseiro) [inline frontend/src/components/Teams/CreateTeam/CreateTeam.tsx:26]:
```suggestion
  const {back} = useRouter();
```

File: frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx [hunk 5]
```diff
@@ -1,57 +1,82 @@
...
+
   const { deleteTeam, deleteTeamUser } = useTeam();
 
-  const { id: userId, isSAdmin } = { ...session?.user };
-  const { id, users, name } = team;
-  const userFound = users.find((member) => member.user?._id === userId);
+  const userFound = useMemo(() => {
+    const queryUserId = router.query.userId;
```
Comment (nunocaseiro) [inline frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx:39]:
```suggestion
    const queryUserId = userPathId;
```

File: frontend/src/hooks/useTeam.tsx [hunk 6]
```diff
@@ -146,6 +148,8 @@ const useTeam = ({
         content: 'The team was successfully created.',
         type: ToastStateEnum.SUCCESS,
       });
+
+      router.push(ROUTES.Teams);
```
Comment (nunocaseiro) [inline frontend/src/hooks/useTeam.tsx:152]:
```suggestion
      push(ROUTES.Teams);
```

File: frontend/src/hooks/useTeam.tsx [hunk 7]
```diff
@@ -46,6 +47,7 @@ const useTeam = ({
     usersList,
     userId,
     session,
+    router,
```
Comment (nunocaseiro) [inline frontend/src/hooks/useTeam.tsx:50]:
```suggestion
    router: {push},
```

File: frontend/src/pages/teams/[teamId].tsx [hunk 8]
```diff
@@ -17,10 +16,15 @@ import Dots from '@/components/Primitives/Loading/Dots/Dots';
...
+import useCurrentSession from '@/hooks/useCurrentSession';
+import { UserList } from '@/types/team/userList';
+import { ROUTES } from '@/utils/routes';
+import { useRouter } from 'next/router';
 
 const Team = () => {
   // Session Details
-  const { data: session } = useSession({ required: true });
+  const { userId, isSAdmin } = useCurrentSession();
+  const router = useRouter();
```
Comment (nunocaseiro) [inline frontend/src/pages/teams/[teamId].tsx:27]:
```suggestion
  const { replace } = useRouter();
```

File: frontend/src/pages/teams/[teamId].tsx [hunk 9]
```diff
@@ -52,13 +60,10 @@ const Team = () => {
...
 
-  const { id: userId, isSAdmin } = { ...session?.user };
-
-  const user = teamData?.users.find((member) => member.user?._id === userId);
-  const hasPermissions =
-    isSAdmin || [TeamUserRoles.ADMIN, TeamUserRoles.STAKEHOLDER].includes(user?.role!);
-
-  if (!session || !teamData || !usersData) return null;
+  if (!teamData || !usersData) {
+    router.replace(ROUTES.Teams);
```
Comment (nunocaseiro) [inline frontend/src/pages/teams/[teamId].tsx:64]:
```suggestion
    replace(ROUTES.Teams);
```

**Review by nunocaseiro — CHANGES_REQUESTED** [review]

File: frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx (hunk 2 above)
Comment (JoaoSaIvador) [reply frontend/src/components/Teams/TeamsList/TeamItem/TeamItem.tsx:33]: In this file, there's another use of the router and it's inside a useMemo (line 37), so I don't think the destructuring would help much in this case.

``` const router = useRouter();
  const isTeamPage = router.pathname.includes('teams');

  const { deleteTeam, deleteTeamUser } = useTeam();

  const userFound = useMemo(() => {
    const queryUserId = router.query.userId;
    const teamUserId = !isTeamPage && queryUserId ? queryUserId : userId;

    return teamUsers.find((teamUser) => String(teamUser.user?._id) === String(teamUserId));
  }, [router, userId, teamUsers]);

**Review by nunocaseiro — APPROVED** [review]
LGTM 🚀

**Pull request merged**